cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java

```diff
@@ -15,44 +15,260 @@
+        updateGroupInstances(groupId, instances, getStartRowsFromColumns(filterByGroup(columns, groupId)));
+      }
+    }
+  }
+
+  /**
+   * Fetches start row states of all instances for a given consumer group.
+   *
+   * @param groupId consumer group Id
+   * @return a map from instanceId to start row state
+   */
+  private Map<Integer, byte[]> fetchStartRows(long groupId) {
+    // Gets all columns with the groupId as prefix
+    Row row = table.get(queueName.toBytes(),
+                        getConsumerStateColumn(groupId, 0),
+                        getConsumerStateColumn(groupId, Integer.MAX_VALUE), Integer.MAX_VALUE);
+    return getStartRowsFromColumns(row.getColumns());
+  }
+
+  /**
+   * Decodes start row states from the given column map.
+   *
+   * @return a Map from instanceId to start row
+   */
+  private Map<Integer, byte[]> getStartRowsFromColumns(Map<byte[], byte[]> columns) {
+    ImmutableMap.Builder<Integer, byte[]> startRows = ImmutableMap.builder();
+    for (Map.Entry<byte[], byte[]> entry : columns.entrySet()) {
+      startRows.put(getInstanceIdFromColumn(entry.getKey()), entry.getValue());
+    }
+    return startRows.build();
+  }
+
+  private Map<byte[], byte[]> filterByGroup(Map<byte[], byte[]> columns, final long groupId) {
+    return Maps.filterKeys(columns, new Predicate<byte[]>() {
+      @Override
+      public boolean apply(byte[] column) {
+        return groupId == getGroupIdFromColumn(column);
+      }
+    });
+  }
+
+  private long getGroupIdFromColumn(byte[] column) {
+    return Bytes.toLong(column);
+  }
+
+  private int getInstanceIdFromColumn(byte[] column) {
+    return Bytes.toInt(column, Bytes.SIZEOF_LONG);
+  }
+
+  /**
+   * Returns the column qualifier for the consumer state column. The qualifier is formed by
+   * {@code <groupId><instanceId>}.
+   * @param groupId Group ID of the consumer
+   * @param instanceId Instance ID of the consumer
+   * @return A new byte[] which is the column qualifier.
+   */
+  private byte[] getConsumerStateColumn(long groupId, int instanceId) {
+    byte[] column = new byte[Longs.BYTES + Ints.BYTES];
+    Bytes.putLong(column, 0, groupId);
+    Bytes.putInt(column, Longs.BYTES, instanceId);
+    return column;
+  }
+
+  /**
+   * Update number of instances for the given group.
+   *
+   * @param groupId consumer group Id
+   * @param instances number of consumer instances in the group
+   * @param startRows existing consumers start row states
+   */
+  private void updateGroupInstances(long groupId, int instances, Map<Integer, byte[]> startRows) {
+    // Find smallest startRow among existing instances
+    byte[] smallest = null;
+    for (byte[] startRow : startRows.values()) {
+      if (smallest == null || Bytes.BYTES_COMPARATOR.compare(startRow, smallest) < 0) {
+        smallest = startRow;
+      }
+    }
+    Preconditions.checkArgument(smallest != null, "No startRow found for consumer group %s", groupId);
```

Comment:
would this happen if one deploys a flow, then updates the instances before starting the flow for the first time? If so, then we should support that case.